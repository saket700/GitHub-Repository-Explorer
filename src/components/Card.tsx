import React from "react";
import { GithubIcon, StarIcon, LicenseIcon, ForkIcon } from "../Icons";

export interface ICardDetails {
  githubUsername: string;
  githubDescription: string | undefined;
  numberOfStars?: string;
  githubLanguage: string;
  visibility: string;
  stargazersCount: number;
  updatedAt: string;
  repoName: string;
  license: {
    key: string;
    name: string;
  };
  forkCount: number;
}
export const Card: React.FC<ICardDetails> = ({
  githubUsername,
  githubDescription,
  githubLanguage,
  visibility,
  stargazersCount,
  updatedAt,
  repoName,
  license,
  forkCount,
}) => {
  return (
    <div className="max-w-[52rem] w-full p-6 bg-white border border-gray-200 rounded-lg shadow ">
      <div className="flex">
        <GithubIcon />
        <a
          href={`/detail?username=${githubUsername}&reponame=${repoName}`}
          target="_blank"
          className="mb-3 text-blue-500 hover:underline"
        >
          {repoName}
        </a>
        <div className="border text-sm text-gray-500 rounded-3xl w-16 h-[24px] mt--[3px] text-uppercase capitalize ml-3 mb-1">
          {visibility}
        </div>
      </div>
      <div className="flex items-start flex-wrap">
        <p className="mb-3 font-normal text-gray-500 dark:text-gray-400 text-sm text-justify">
          {githubDescription ? githubDescription : "-"}
        </p>
      </div>
      <div className="flex pt-2 text-xs">
        {!!stargazersCount && (
          <>
            <p className="mb-3 font-normal text-gray-500 dark:text-gray-400 mr-3 pt-1">
              {githubLanguage}
            </p>
            <StarIcon />
            <div className="font-normal  text-gray-400 pt-1">{stargazersCount}</div>
          </>
        )}
        {!!forkCount && (
          <div className=" flex pl-2">
            <div className="pt-1">
              <ForkIcon />
            </div>
            <div className="font-normal  text-gray-500 dark:text-gray-400 mr-3 pt-1">
              {forkCount}
            </div>
          </div>
        )}
        {!!license?.name && (
          <div className=" flex pl-2">
            <div className="pt-1">
              <LicenseIcon />
            </div>
            <div className="font-normal  text-gray-500 dark:text-gray-400 mr-3 pt-1">
              {license?.name}
            </div>
          </div>
        )}
        <div className="text-gray-400 font-normal text-xs pl-2 pt-1">
          Updated on {updatedAt}
        </div>
      </div>
    </div>
  );
};
