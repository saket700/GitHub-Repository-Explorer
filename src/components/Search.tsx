import React, { useState } from "react";
import { SearchIcon } from "../Icons";

interface ISearch {
  onSearchClick: (searchValue: string) => void;
}

export const Search: React.FC<ISearch> = ({ onSearchClick: onSearchClick }) => {
  const [searchValue, setSearchValue] = useState("");

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    onSearchClick(searchValue);
  };

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearchValue(e.target.value);
  };

  return (
    <div className="pt-4 max-w-2xl w-full">
      <form onSubmit={handleSubmit}>
        <label
          htmlFor="default-search"
          className="mb-2 text-sm font-medium text-gray-900 sr-only dark:text-white"
        >
          Search
        </label>
        <div className="relative hover:bg-transparent hover:border-transparent hover:text-white">
          <input
          aria-label="search"
            id="username"
            value={searchValue}
            onChange={handleInputChange}
            className="block w-full p-2 ps-3 text-sm text-gray-900 border border-gray-300 rounded-lg bg-gray-50  dark:placeholder-gray-400"
            placeholder="Find a repository"
            required
          />
          <button
            data-testid="Search"
            type="submit"
            className="text-white absolute end-2.5 bottom-1.5 font-medium rounded-lg text-sm px-4 py-1 pt-[20px]"
          >
            <div className="absolute inset-y-0 start-0 flex items-center ps-3 pointer-events-none ">
              <SearchIcon />
            </div>
          </button>
        </div>
      </form>
    </div>
  );
};
