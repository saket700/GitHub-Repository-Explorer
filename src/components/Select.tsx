import React from "react";

type OptionValue = string;

interface SelectProps<T extends OptionValue> {
  options: T[];
  value: T;
  onChange: (event: React.ChangeEvent<HTMLSelectElement>) => void;
  isDisabled?: boolean;
}

export const Select: React.FC<SelectProps<OptionValue>> = ({ onChange, isDisabled,options }) => {
  return (
    <div className="ml-4 mt-[16px]">
      <select
        id="github_repo"
        className="bg-gray-50 border border-gray-300 text-sm rounded-lg block w-full p-4 text-black bottom-1.5 end-2.5 py-[10px]"
        onChange={onChange} 
        disabled={isDisabled}
        
      >
        {options.map((option) => (
          <option key={option} value={option} >
            {option}
          </option>
        ))}
      </select>
    </div>
  );
};
