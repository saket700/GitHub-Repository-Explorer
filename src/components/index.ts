export * from './Card';
export * from './CardLayout';
export * from './Loader';
export * from './NotFound';
export * from './Pagination';
export * from './Search';
export * from './Select';
