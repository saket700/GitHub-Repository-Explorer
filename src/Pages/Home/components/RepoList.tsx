import React from "react";
import { IGithubUserDetails } from "../../../types/GithubUserDetailsType";
import { formatDate } from "../../../utils";
import { Card } from "../../../components/Card";

interface IRepoList {
  currentPosts: IGithubUserDetails[];
}
export const RepoList: React.FC<IRepoList> = ({ currentPosts }) => (
  <div className="pt-8">
    {currentPosts?.map((item: IGithubUserDetails) => (
      <div key={item.id}  className="mb-8">
        <Card
          
          githubUsername={item?.owner?.login}
          githubDescription={item?.description}
          numberOfStars={item.stargazers_url}
          repoName={item.name || ""}
          githubLanguage={item?.language}
          visibility={item?.visibility}
          stargazersCount={item?.stargazers_count}
          updatedAt={formatDate(item.updated_at, {
            month: "long",
            day: "numeric",
            year: "numeric"
          })}
          license={item?.license}
          forkCount={item.forks_count}
        />
      </div>
    ))}
  </div>
);
