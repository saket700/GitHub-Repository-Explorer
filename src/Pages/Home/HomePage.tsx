import React, { useState } from "react";
import { paginate, sortRepos, useGitHubRepos } from "../../utils";
import { Select, Loader, NotFound, Search, Pagination } from '../../components'
import { RepoList } from "./components/RepoList";


const pageSize = 5;
const updatedData = "Last updated";
const OptionList = [
  { id: 1, label: "Last updated" },
  { id: 2, label: "Name" },
  { id: 3, label: "Stars" },
];
export const HomePage = () => {
  const [searchValue, setSearchValue] = useState("");
  const [currentPage, setCurrentPage] = useState(1);
  const [sortOption, setSortOption] = useState(updatedData);

  const { repos, loading, fetchGitHubRepos } = useGitHubRepos();
  const sortedRepos = sortRepos(repos, sortOption);

  const onSearch = (value: string) => {
    setSearchValue(value);
    fetchGitHubRepos(value);
    setCurrentPage(1);
  };

  const handleSortChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
    setSortOption(event.target.value);
  };

  const currentPosts = paginate(sortedRepos, currentPage, pageSize);

  return (
    <div className="m-auto max-w-[900px]">
      <label
        htmlFor="github_username"
        className="text-sm font-medium leading-6 text-gray-900 flex justify-center"
      >
        Github Explorer Repository
      </label>
      <div className="flex flex-col md:flex-row md:items-center">
        <Search onSearchClick={onSearch} />
        <Select
          onChange={handleSortChange}
          isDisabled={!searchValue}
          options={OptionList.map((option) => option.label)}
          value={sortOption}
        />
      </div>
      {loading && (
        <div className="flex justify-center pt-10">
          <Loader />
        </div>
      )}

      {!loading && searchValue && currentPosts.length===0 && (
        <div className="pt-16">
          <NotFound
            headingNotFound={"Repository not found"}
            descriptionNotFound={`We are sorry but ${searchValue} doesnâ€™t have any repositories that match.`}
          />
        </div>
      )}

      <RepoList currentPosts={currentPosts} />

      <div className="w-full flex justify-center">
        <Pagination
          currentPage={currentPage}
          totalCount={repos.length}
          pageSize={pageSize}
          onPageChange={(page) => setCurrentPage(page)}
          siblingCount={0}
        />
      </div>
    </div>
  );
};

export default HomePage;
