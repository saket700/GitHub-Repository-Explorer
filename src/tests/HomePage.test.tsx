import {
  fireEvent,
  render,
  waitFor,
  screen,
} from "@testing-library/react";
import HomePage from "../Pages/Home/HomePage";
import renderer from "react-test-renderer";
import React from "react";
import { vi } from "vitest";
import { mockData } from "../mock-data/mockData";
import { Pagination } from "../components";


describe("Snapshot testing suite", () => {
  it("Matches DOM Snapshot", () => {
    const domTree = renderer.create(<HomePage />).toJSON();
    expect(domTree).toMatchSnapshot();
  });
});


test("renders SearchComponent & select component", () => {
  render(<HomePage />);
  expect(screen.getByText("Github Explorer Repository")).toBeInTheDocument();
  expect(screen.getByLabelText("search")).toBeInTheDocument();
  expect(screen.getByRole("combobox")).toBeInTheDocument();
});

test("renders Select component and is disabled initially", () => {
  render(<HomePage />);
  const select = screen.getByRole("combobox");
  expect(select).toBeInTheDocument();
  expect(select).toBeDisabled();
});

test("changing sorting option updates state and re-renders", () => {
  render(<HomePage />);
  const select = screen.getByRole("combobox") as HTMLInputElement;
  fireEvent.change(select, { target: { value: "Last updated" } });
  expect(select.value).toBe("Last updated");
});



test("render data on search", async () => {
  vi.mock("fetch", () => ({
    ok: true,
    json: () =>
      Promise.resolve({
        success: true,
        status: 200,
        data: mockData,
      }),
  }));
  render(<HomePage />);
  fireEvent.change(screen.getByLabelText("search"), {
    target: { value: "ivey" },
  });

  const searchButton = screen.getByTestId("Search");
  fireEvent.click(searchButton);

  await waitFor(() => {
    const repoNameElement = screen.getByText("ivey.github.com");
    expect(repoNameElement).toBeInTheDocument();
  });
  const select = screen.getByRole("combobox") as HTMLInputElement;
  fireEvent.change(select, { target: { value: "Stars" } });
  expect(select.value).toBe("Stars");
  await waitFor(() => {
    const repoNameElement = screen.getByText("rss-to-notifo");
    expect(repoNameElement).toBeInTheDocument();
  });
});

test("render no data if no repository found", async () => {
  vi.mock("fetch", () => ({
    ok: true,
    json: () =>
      Promise.resolve({
        success: true,
        status: 200,
        data: mockData,
      }),
  }));
  render(<HomePage />);
  fireEvent.change(screen.getByLabelText("search"), {
    target: { value: "asd" },
  });

  const searchButton = screen.getByTestId("Search");
  fireEvent.click(searchButton);

  await waitFor(() => {
    const repoNameElement = screen.getByText("Repository not found");
    expect(repoNameElement).toBeInTheDocument();
  });
});

test("should render the pagination buttons ", () => {
  const onPageChangeMock = vi.fn();
  const component = render(
    <Pagination
      onPageChange={onPageChangeMock}
      totalCount={10}
      siblingCount={1}
      currentPage={1}
      pageSize={5}
    />
  );

  const previousButton = screen.queryByText("Previous");
  const nextButton = screen.queryByText("Next");

  expect(previousButton).toBeInTheDocument();
  expect(nextButton).toBeInTheDocument();

  // Find and click the "Next" button
  const nextButtonClick = screen.getByText("Next");
  fireEvent.click(nextButtonClick);

  expect(onPageChangeMock).toHaveBeenCalledWith(2);
});
