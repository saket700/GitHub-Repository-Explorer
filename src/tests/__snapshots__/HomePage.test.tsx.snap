// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot testing suite > Matches DOM Snapshot 1`] = `
<div
  className="m-auto max-w-[900px]"
>
  <label
    className="text-sm font-medium leading-6 text-gray-900 flex justify-center"
    htmlFor="github_username"
  >
    Github Explorer Repository
  </label>
  <div
    className="flex flex-col md:flex-row md:items-center"
  >
    <div
      className="pt-4 max-w-2xl w-full"
    >
      <form
        onSubmit={[Function]}
      >
        <label
          className="mb-2 text-sm font-medium text-gray-900 sr-only dark:text-white"
          htmlFor="default-search"
        >
          Search
        </label>
        <div
          className="relative hover:bg-transparent hover:border-transparent hover:text-white"
        >
          <input
            aria-label="search"
            className="block w-full p-2 ps-3 text-sm text-gray-900 border border-gray-300 rounded-lg bg-gray-50  dark:placeholder-gray-400"
            id="username"
            onChange={[Function]}
            placeholder="Find a repository"
            required={true}
            value=""
          />
          <button
            className="text-white absolute end-2.5 bottom-1.5 font-medium rounded-lg text-sm px-4 py-1 pt-[20px]"
            data-testid="Search"
            type="submit"
          >
            <div
              className="absolute inset-y-0 start-0 flex items-center ps-3 pointer-events-none "
            >
              <svg
                aria-hidden="true"
                className="w-4 h-4 text-gray-500 dark:text-gray-400"
                fill="none"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m19 19-4-4m0-7A7 7 0 1 1 1 8a7 7 0 0 1 14 0Z"
                  stroke="currentColor"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                />
              </svg>
            </div>
          </button>
        </div>
      </form>
    </div>
    <div
      className="ml-4 mt-[16px]"
    >
      <select
        className="bg-gray-50 border border-gray-300 text-sm rounded-lg block w-full p-4 text-black bottom-1.5 end-2.5 py-[10px]"
        disabled={true}
        id="github_repo"
        onChange={[Function]}
      >
        <option
          value="Last updated"
        >
          Last updated
        </option>
        <option
          value="Name"
        >
          Name
        </option>
        <option
          value="Stars"
        >
          Stars
        </option>
      </select>
    </div>
  </div>
  <div
    className="pt-8"
  />
  <div
    className="w-full flex justify-center"
  />
</div>
`;
